Template/Trigger/ElementVisibilityTrigger.web.js

```diff
@@ -235,9 +230,82 @@
             };
         }
 
+        function getSelectors() {
+            var selectionMethod = parameters.get('selectionMethod');
+            if (selectionMethod === 'elementId') {
+                return '#' + parameters.get('elementId');
+            } else if (selectionMethod === 'cssSelector') {
+                return parameters.get('cssSelector');
+            }
+
+            return;
+        }
+
+        function setIntersectionObserver(triggerEvent) {
+            if ('IntersectionObserver' in window) {
+                var interSectionObserverOptions = {
+                    root: null, // document's viewport as the container.
+                    rootMargin: '0px',
+                    threshold: (minPercentVisible / 100)
+                };
+                var observerIntersection = new IntersectionObserver(function (entries) {
+                    interSectionCallback(entries, triggerEvent, observerIntersection);
+                }, interSectionObserverOptions);
+
+                if (selectors) {
+                    TagManager.dom.bySelector(selectors).forEach(function (element) {
+                        observerIntersection.observe(element);
+                    });
+                }
+            }
+
+        }
+
+        function interSectionCallback(entries, triggerEvent, observer) {
+            var dom = TagManager.dom;
+            entries.forEach(function (entry) {
+                if (entry.intersectionRatio > 0) {
+                    if (blockTrigger || (onlyOncePerElement && isNodeEventTriggered(entry.target))) {
+                        return;
+                    }
+                    triggerEvent({
+                        event: 'mtm.ElementVisibility',
+                        'mtm.elementVisibilityPercentage': minPercentVisible,
+                        'mtm.elementVisibilityId': dom.getElementAttribute(entry.target, 'id'),
+                        'mtm.elementVisibilityClasses': dom.getElementClassNames(entry.target),
+                        'mtm.elementVisibilityText': TagManager.utils.trim(entry.target.innerText),
+                        'mtm.elementVisibilityNodeName': entry.target.nodeName,
+                        'mtm.elementVisibilityUrl': entry.target.href || dom.getElementAttribute(entry.target, 'href'),
+                    });
+
+                    if (fireTriggerWhen === 'oncePage') {
+                        blockTrigger = true;
+                        TagManager.dom.bySelector(selectors).forEach(function (element) {
+                            observer.unobserve(element);
```

Comment:
would we check here and below if there is maybe `self.scrollIndex` defined and in this case also call `
                                TagManager.window.offScroll(self.scrollIndex);`?

Reply:
and where we call `TagManager.window.offScroll(self.scrollIndex);` would we maybe call `observer.unobserve(element);`?

Reply:
Yes added both at the respective callbacks